File: dynamic-programming/70-climbing-stairs.py
class Solution:
    def __init__(self):
        self.memo = []

    def helper(self, n: int) -> int:
        if n == 2:
            return 2
        elif n == 1:
            return 1
        elif n == 0:
            return 0

        if self.memo[n] > 0:
            return self.memo[n]

        self.memo[n-1] = self.helper(n-1)
        self.memo[n-2] = self.helper(n-2)

        return self.memo[n-1] + self.memo[n-2]

    def climbStairs(self, n: int) -> int:
        self.memo = [0] * (n + 1)
        return self.helper(n)


print(Solution().climbStairs(2))
print(Solution().climbStairs(3))
print(Solution().climbStairs(4))

File: easy/dp/cpp/70-climbing-stairs.cpp
#include "../../../stdc++.h"
using std::unordered_map;

class Solution
{
private:
    unordered_map<int, int> memo;

public:
    int climbStairs(int n)
    {
        for (int i = 0; i <= n; i++)
        {
            memo[i] = -1;
        }
        memo[0] = 0;
        memo[1] = 1;
        memo[2] = 2;

        return helper(n);
    }

    int helper(int n)
    {
        if (memo[n] != -1)
        {
            return memo[n];
        }
        else
        {
            memo[n] = helper(n - 1) + helper(n - 2);
        }
        return memo[n];
    }
};

void test(int n)
{
    Solution a;
    int result = a.climbStairs(n);
    std::cout << result << std::endl;
}

int main()
{
    // test(1);
    test(2);
    test(3);
    test(4);
}
